packages/nimble-components/src/select/index.ts

```diff
@@ -969,39 +984,89 @@ export class Select
+
+    private filterMatchesText(text: string): boolean {
+        const filter = this.filter.toLowerCase();
+        const normalizedFilter = diacriticInsensitiveStringNormalizer(filter);
+        return diacriticInsensitiveStringNormalizer(text).includes(
+            normalizedFilter
+        );
+    }
+
     /**
      * Filter available options by text value.
      *
      * @public
      */
     private filterOptions(): void {
-        const filter = this.filter.toLowerCase();
-
-        if (filter) {
-            this.filteredOptions = this.options.filter(option => {
-                const normalizedFilter = diacriticInsensitiveStringNormalizer(filter);
-                return (
-                    !option.hidden
-                    && diacriticInsensitiveStringNormalizer(option.text).includes(
-                        normalizedFilter
-                    )
-                );
-            });
-        } else {
-            this.filteredOptions = this.options.filter(
-                option => !option.hidden
-            );
+        if (!this.$fastController.isConnected) {
+            return;
         }
 
-        this.options.forEach(o => {
-            if (isNimbleListOption(o)) {
-                if (!this.filteredOptions.includes(o)) {
-                    o.visuallyHidden = true;
-                } else {
-                    o.visuallyHidden = false;
+        const filteredOptions: ListOption[] = [];
+
+        let lastVisibleOptionGroupIndex = -1;
+        let lastVisibleElement: ListOption | ListOptionGroup | undefined;
+        for (let i = 0; i < this.slottedOptions.length; i++) {
+            const element = this.slottedOptions[i]!;
+
+            if (element instanceof ListOptionGroup) {
+                element.showTopSeparator = false;
+                element.showBottomSeparator = false;
+                const groupOptions = this.getGroupOptions(element);
+                const groupMatchesFilter = this.filterMatchesText(
+                    element.labelContent
+                );
+                let allOptionsHidden = true;
+                groupOptions.forEach(option => {
+                    option.visuallyHidden = groupMatchesFilter
+                        ? false
+                        : this.isOptionHidden(option);
+                    if (!option.visuallyHidden) {
+                        filteredOptions.push(option);
+                    }
+                    allOptionsHidden = option.visuallyHidden && allOptionsHidden;
+                });
+                element.visuallyHidden = allOptionsHidden || element.hidden;
+
+                if (!allOptionsHidden) {
+                    element.showBottomSeparator = true;
+                    lastVisibleOptionGroupIndex = i;
+                    element.showTopSeparator = lastVisibleElement instanceof ListOption;
+                }
+                lastVisibleElement = allOptionsHidden
+                    ? lastVisibleElement
+                    : element;
```

Comment:
Instead of writing this logic in both if-statements, can you write it a single time for both?
```ts
for (...) {
    if (isGroup) {
       ...
    } else if (isOption) {
       ...
    }

    if (!element.visuallyHidden) {
        lastVisibleElement = element;
    }
}
```